### PR TITLE
Add Referee page, with authentication token

### PR DESCRIPTION
**Pull request by alexdesi**
## Context

This PR is the first slice of replacing the google form to a Referee page. This includes:

- Create the token for the referee (stored on Reference model)
- Create a new page to allow the referee to enter the comments https://apply-beta-prototype.herokuapp.com/reference/comments
- Include the token in the email and check the token is correct when trying to access the reference view

## Changes proposed in this pull request

- new Token on `Reference` model
- update email with new Token
- Add new view/controller in a new namespace
- Added new feature Flag `reference_form`

## Guidance to review
Tests are a good starting point

## Link to Trello card

https://trello.com/c/RrlqYDMz/585-referee-can-provide-a-reference-via-dfe-apply-service-%F0%9F%8F%88

### CODE REVIEW

**Comment by alexdesi** [comment]
Screenshot 2019-12-12 at 17 00 30

File: app/services/submit_application.rb [hunk 1]
```diff
@@ -31,5 +31,10 @@ def submit_application
       application_choice.edit_by = ApplicationDates.new(application_form).edit_by
       ApplicationStateChange.new(application_choice).submit!
     end
+
+    application_form.references.includes(:application_form).each do |ref|
```
Comment (tijmenb) [inline app/services/submit_application.rb]: Perhaps we should do this in the same loop as the mail is sent, because the ordering is important.

Comment (alexdesi) [reply app/services/submit_application.rb]: Yes, it's nice to have the update at the same time the email is sent.

I've just one small concern: `send_reference_request_email_to_referees` will do 2 different things, instead of just one. (but I think it's fine moving the update_token code in the model, have a look)

File: app/controllers/referee_interface/reference_controller.rb [hunk 2]
```diff
@@ -0,0 +1,47 @@
+module RefereeInterface
+  class ReferenceController < ActionController::Base
+    layout 'application'
```
Comment (tijmenb) [inline app/controllers/referee_interface/reference_controller.rb]: We'll need to add some things here that the other controllers have as well:

- `include LogQueryParams`
- An implementation of `before_action :add_identity_to_log` so that we know in the logs & Sentry who triggered a log or error

See the candidate bit for an example:

https://github.com/DFE-Digital/apply-for-postgraduate-teacher-training/blob/d4010b0622ac6d86239210ca8197b72acbcc941a/app/controllers/candidate_interface/candidate_interface_controller.rb#L3-L7

File: app/controllers/referee_interface/reference_controller.rb [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+    def feedback
+      reference = Reference.find_by(token: params[:token])
```
Comment (tijmenb) [inline app/controllers/referee_interface/reference_controller.rb]: If you do `Reference.find_by!` it'll trigger an `RecordNotFound` exception that you can catch with a `rescue_from`.

File: app/controllers/referee_interface/reference_controller.rb [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+      if reference.present?
+        @reference_form = RefereeInterface::ReferenceForm.build_from_reference(reference)
+        @application = reference.application_form
+      else
+        render_404
+      end
+    end
+
+    def submit_feedback
+      reference = Reference.find_by(token: params[:token])
```
Comment (tijmenb) [inline app/controllers/referee_interface/reference_controller.rb]: This doesn't check if the token exists before doing the work, using `find_by!` would help. 

I also wonder what happens if the token is empty, since it'll match the `Reference` that currently don't have a `token` set.

File: app/helpers/application_helper.rb [hunk 5]
```diff
@@ -36,4 +36,8 @@ def service_link
   def current_namespace
     params[:controller].split('/').first
   end
+
+  def full_name(application_form)
```
Comment (tijmenb) [inline app/helpers/application_helper.rb]: This probably warrants to be put in `ApplicationForm`.

File: app/views/referee_interface/reference/finish.html.erb [hunk 6]
```diff
@@ -0,0 +1,7 @@
+<main class="govuk-main-wrapper govuk-main-wrapper--auto-spacing" id="main-content" role="main">
+  <h1 class="govuk-heading-xl">Thank you</h1>
+    <div class="govuk-grid-row">
```
Comment (tijmenb) [inline app/views/referee_interface/reference/finish.html.erb]: Minor: some weird indentation here.

File: app/views/referee_mailer/reference_request_email.text.erb [hunk 7]
```diff
@@ -1,20 +1,10 @@
...
-<% end %>
-
-Please use the link below to give a complete reference as soon as possible.
-
-<%= @google_form_url %>
-
-If we don’t hear from you within 5 working days we’ll send you a reminder.
-
-If you won’t give <%= @candidate_name %> a reference, please let us know as soon as possible by emailing: [becomingateacher@digital.education.gov.uk](mailto:becomingateacher@digital.education.gov.uk)
+<%= @reference_link %>
```
Comment (tijmenb) [inline app/views/referee_mailer/reference_request_email.text.erb]: Do we not need to add the paragraph describing what to do when they don't want the reference here?

File: config/environments/test.rb [hunk 8]
```diff
@@ -39,7 +39,7 @@
   config.action_mailer.default_options = {
     from: 'mail@example.com',
   }
-  config.action_mailer.default_url_options = { host: 'localhost', port: 3000 }
+  config.action_mailer.default_url_options = { host: 'www.example.com', port: 3000 }
```
Comment (tijmenb) [inline config/environments/test.rb]: What's the reason behind this change?

File: config/routes.rb [hunk 9]
```diff
@@ -211,6 +211,13 @@
     end
   end
 
+  namespace :referee_interface, path: '/reference' do
+    get '/finish' => 'reference#finish', as: :reference_finish
+    get '/:token/confirmation' => 'reference#confirmation', as: :confirmation
+    get '/:token' => 'reference#feedback', as: :reference_feedback
+    patch '/:token/confirmation' => 'reference#submit_feedback', as: :submit_feedback
+  end
```
Comment (tijmenb) [inline config/routes.rb]: By default we would log the full URL, including the token. If we'd added the token in the query param it would be automatically filtered out.

File: app/views/referee_interface/reference/feedback.html.erb [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+            <p class="govuk-body">For example, can the candidate meet deadlines and manage their time?</p>
+          </li>
+        </ul>
+
+        <div class="govuk-character-count" data-module="govuk-character-count" data-maxwords="300">
+          <div class="govuk-form-group">
+            <label class="govuk-label govuk-label--m govuk-!-margin-top-6" for="reference-comments">
+              Your reference
+            </label>
+            <textarea class="govuk-textarea govuk-js-character-count" id="reference-comments" name="reference[comments]" rows="15" aria-describedby="reference-comments-info"></textarea><span id="reference-comments-info" class="govuk-hint govuk-character-count__message" aria-live="polite">You have 300 words remaining</span>
```
Comment (tijmenb) [inline app/views/referee_interface/reference/feedback.html.erb]: Any reason not to use the formbuilder here?

File: app/models/reference.rb [hunk 11]
```diff
@@ -29,4 +29,9 @@ def email_address_not_own
 
     errors.add(:email_address, :own) if email_address == candidate_email_address
   end
+
+  def update_token
+    raw_token, _encrypted_token = Devise.token_generator.generate(Reference, :token)
+    update(token: raw_token)
```
Comment (tijmenb) [inline app/models/reference.rb]: I would've expected the `raw_token` to be sent to the user, and only the encrypted token saved in the database. 

This is how `VendorApiToken` and `Candidate` tokens work - it prevents the tokens from being used if they ever leak. Open to discussion here!

File: app/views/referee_interface/reference/confirmation.html.erb [hunk 12]
```diff
@@ -0,0 +1,11 @@
+<main class="govuk-main-wrapper govuk-main-wrapper--auto-spacing" id="main-content" role="main">
+
+  <%= form_with model: @reference, url: referee_interface_reference_finish_path, method: :get do |f| %>
```
Comment (tijmenb) [inline app/views/referee_interface/reference/confirmation.html.erb]: This doesn't need to be a form I think?

File: app/views/referee_interface/reference/feedback.html.erb [hunk 13]
```diff
@@ -0,0 +1,49 @@
+<% content_for :title, t('page_titles.application') %>
+
+<main class="govuk-main-wrapper govuk-main-wrapper--auto-spacing" id="main-content" role="main">
+
+  <%= form_with model: @reference_form, url: referee_interface_submit_feedback_path, method: :patch do |f| %>
```
Comment (tijmenb) [inline app/views/referee_interface/reference/feedback.html.erb]: I'm surprised the token is automatically passed into this URL, didn't know Rails did that!

File: app/models/reference.rb (hunk 11 above)
Comment (tijmenb) [inline app/models/reference.rb]: `update!` to avoid silent failures (we're not checking the return value).

```suggestion
    update!(token: raw_token)
```

File: app/controllers/referee_interface/reference_controller.rb [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+
+      if reference.present?
+        @reference_form = RefereeInterface::ReferenceForm.build_from_reference(reference)
+        @application = reference.application_form
+      else
+        render_404
+      end
+    end
+
+    def submit_feedback
```
Comment (tijmenb) [inline app/controllers/referee_interface/reference_controller.rb]: We're going to have to check if the feedback has already been given, otherwise we're allowing edits forever, which might lead to weirdness.

File: app/models/referee_interface/reference_form.rb [hunk 15]
```diff
@@ -0,0 +1,24 @@
+module RefereeInterface
+  class ReferenceForm
```
Comment (tijmenb) [inline app/models/referee_interface/reference_form.rb]: We should use (a version of / refactored version of) `ReceiveReference` instead of this new form object. 

In the flow, this makes sure that when all references are complete the state is changed and we send the applications to the providers.

File: app/views/referee_mailer/reference_request_email.text.erb (hunk 7 above)
Comment (C-gyorfi) [reply app/views/referee_mailer/reference_request_email.text.erb]: Yes, I resolved this in 4354366

File: app/controllers/referee_interface/reference_controller.rb (hunk 14 above)
Comment (alexdesi) [reply app/controllers/referee_interface/reference_controller.rb]: Good point, I've added in my todo list.

File: app/helpers/application_helper.rb (hunk 5 above)
Comment (alexdesi) [reply app/helpers/application_helper.rb]: it sounds good (possible we'll need a Container for that view, so I might put full_name in it)

File: app/models/reference.rb (hunk 11 above)
Comment (alexdesi) [reply app/models/reference.rb]: Yes, I should have added a big `TODO` there, it needs to be secured.

File: app/views/referee_interface/reference/confirmation.html.erb (hunk 12 above)
Comment (alexdesi) [reply app/views/referee_interface/reference/confirmation.html.erb]: That's because the radio button for the consent to be contacted is coming soon,
so it'll need to be a form.

File: app/views/referee_interface/reference/feedback.html.erb (hunk 13 above)
Comment (alexdesi) [reply app/views/referee_interface/reference/feedback.html.erb]: you made me doubt of it :D, I think it is working,  I'll double check

File: app/views/referee_interface/reference/feedback.html.erb (hunk 10 above)
Comment (alexdesi) [reply app/views/referee_interface/reference/feedback.html.erb]: not really, I'll fix it!

File: app/models/reference.rb (hunk 11 above)
Comment (alexdesi) [reply app/models/reference.rb]: Acctually we added that iten to the https://trello.com/c/B4tokVQU/§, because we were aware this PR was getting long.

File: app/views/referee_mailer/reference_request_email.text.erb (hunk 7 above)
Comment (alexdesi) [reply app/views/referee_mailer/reference_request_email.text.erb]: Acctually we added that iten to the https://trello.com/c/B4tokVQU/§, because we were aware this PR was getting long.

File: config/environments/test.rb (hunk 8 above)
Comment (alexdesi) [reply config/environments/test.rb]: That is only of tests, I thought that having to use `localhost:3000` inside the test was a bit weird.
But not strong opinion on that, possible I should keep that out, since this PR is already quite long.

File: app/controllers/referee_interface/reference_controller.rb (hunk 2 above)
Comment (alexdesi) [reply app/controllers/referee_interface/reference_controller.rb]: good point! 
I think we need all of them but `before_action :authenticate_candidate! ` since in this case, the user is the referee

Comment (alexdesi) [reply app/controllers/referee_interface/reference_controller.rb]: For the  `before_action :authenticate_referee!`
I am gonna use the same password as `ENV.fetch('BASIC_AUTH_PASSWORD'))`

I suppose it makes sense since who will test this in staging/production is the same user that creates the application form.

Comment (alexdesi) [reply app/controllers/referee_interface/reference_controller.rb]: I am not sure the data to log with `add_identity_to_log`.
Probably just the `id` of the `referee`.  (like the support controllers, candidate controller, ...)

File: app/models/referee_interface/reference_form.rb (hunk 15 above)
Comment (alexdesi) [reply app/models/referee_interface/reference_form.rb]: As we discussed, it's fine to have a bit of duplication.

File: app/controllers/referee_interface/reference_controller.rb (hunk 14 above)
Comment (alexdesi) [reply app/controllers/referee_interface/reference_controller.rb]: I've added a condition on the `reference_controller#feedback` to check if the feedback already exists.

File: app/models/reference.rb (hunk 11 above)
Comment (alexdesi) [reply app/models/reference.rb]: Done

Comment (alexdesi) [reply app/models/reference.rb]: done

File: app/views/referee_interface/reference/feedback.html.erb (hunk 13 above)
Comment (alexdesi) [reply app/views/referee_interface/reference/feedback.html.erb]: anyway, now it uses a query string `?token=XXXXXX`
